keps/sig-apps/2232-jobs-stopped/README.md

```diff
@@ -0,0 +1,545 @@
+
+The canonical place for the latest set of instructions (and the likely source
+of this file) is [here](/keps/NNNN-kep-template/README.md).
+
+**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
+it is marked `implementable`, must be approved by each of the KEP approvers.
+If none of those approvers are still appropriate, then changes to that list
+should be approved by the remaining approvers and/or the owning SIG (or
+SIG Architecture for cross-cutting KEPs).
+-->
+# KEP-2322: Stopped Jobs
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Constraints/Caveats](#notesconstraintscaveats)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Scalability](#scalability)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed](#infrastructure-needed)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The Jobs API allows users to create Pods using the jobs controller with
+specific requirements such as Pod-level parallelism, completing a certain
+number of executions, and restart policies.
+
+This KEP proposes an enhancement that will allow stopping and resuming Jobs.
```

Comment:
This looks mostly good, my only suggestion would be not to use `stopped` which might suggest we are halting the job, but maybe rather suspend or paused? I'm personally leaning towards the former, since that's what CronJob has so having it consistent for both batch resources would be desirable.

Reply:
I didn't know CronJob had a suspend field. In that case, I prefer suspend too.

@ahg-g @alculquicondor opinions before I make the change?

Reply:
+1 to suspended. pause exist in deployment, but has completely different semantics, so -1 to that.

Reply:
+1 We discussed and agreed to this at the sig meeting.

Reply:
Hmm - I'm actually wondering about that.
In CronJob - we are suspending "subsequent executions":
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/batch/v1beta1/types.go#L119

Here - we are also stopping the currently running ones.
So isn't stopped better reflecting that?
@soltysh @ahg-g - thoughts?

Reply:
I am agnostic, in https://github.com/kubernetes/kubernetes/issues/97170 this I used "suspended/stopped" :)

But I agree that stopped would then be better if the CrongJob semantics doesn't include stopping a currently running job.

Reply:
kept it Stopped for now, pls let me know if this needs to change.

Reply:
Can you add something like the lines of

We will discuss with API approver which of "stopped" or "suspended" is a better name, given existence of "suspended" in CronJob (though with slightly different semantics).

Does that make sense?

Reply:
@lavalamp can you please give your opinion on the flag name?

Reply:
I'll be in favor of suspended, but like @wojtek-t mentioned this can be resolved at api review time and updated here.

Reply:
If you weren't killing the existing pods, I'd vote for "paused" or "suspended".

If you restart a job after it's been stopped, does the whole thing restart or does it count the pods that already did succeed?

If it starts over, then stop/start seems right. If it latches progress, then maybe pause/continue or suspend/resume is better.

Reply:
it will will latch on progress made, finished pods will count towards the completion status of the job.

Reply:
I think "suspend" is a good word, it strongly implies that a "resume" is possible, and also (at least to me) implies more action wrt the currently running pods than "pause" would.

"stop" is probably my least favorite, since stopped things can't always be restarted and also might start from the beginning.

(I think this is a little subjective)

I think it needs to be very clear about what happens to currently running pods, I personally would have guessed that users don't want to lose any progress those pods have made. Are you sure users won't want that?

Reply:
> I think it needs to be very clear about what happens to currently running pods, I personally would have guessed that users don't want to lose any progress those pods have made. Are you sure users won't want that?

This is the same semantics as setting `parallelism=0`. Deleting running pods allows a job-level scheduler to "preempt" the job.

Reply:
OK, that makes sense to me, I think "suspend" is a good term, and I imagine the documentation for such a scheduler will explain the potential consequences. sgtm.

Reply:
The Pods still get a termination signal, so they have a chance to save their work.

Suspend implies some middle ground. I think the relative ambiguity of it will at least trigger users to read the field documentation.

Reply:
@lavalamp The only drawback of suspend I see is the inconsistency with the "CronJob" semantics (where we don't kill the already running job). But if you think that's ok, I'm fine with it.

Reply:
@wojtek-t Yeah, I hear that. I'm somewhat torn between a) just living with it, b) adding subtypes to "suspend" semantics (e.g., "hard" vs "soft" suspension), or c) coming up with a different word--unfortunately in a vacuum I think "suspend" is the best word :/

I lean towards option a) and fixing it with documentation. Maybe we should get a 3rd opinion though.

Reply:
I don't have a better suggestion too (and I think (b) will introduce even more confusion). So I'm fine with (a) too - just wanted to ensure we all are aware of that (and we are) - which makes me happy.

Reply:
(b) isn't a bad idea. It's also possible to extend this field to have soft & hard suspension (hard as default when `suspend` is set) after alpha, so let's go with (a) for now until we gather more user feedback.